### ISSUE TITLE
ARS implementation inverts the priority of servicetime

### ISSUE THREAD

**Issue by robertclancy** [elastic/elasticsearch#65838]
**Elasticsearch version** (`bin/elasticsearch --version`): commit c3ab6f71c873551ebef333c655b348023a497d9b

**Plugins installed**: []

**JVM version** (`java -version`):
```
openjdk version "12.0.2" 2019-07-16
OpenJDK Runtime Environment AdoptOpenJDK (build 12.0.2+10)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 12.0.2+10, mixed mode)
```

**OS version** (`uname -a` if on a Unix-like system):
```
Darwin Roberts-MacBook-Pro.local 19.6.0 Darwin Kernel Version 19.6.0: Thu Oct 29 22:56:45 PDT 2020; root:xnu-6153.141.2.2~1/RELEASE_X86_64 x86_64
```

**Description of the problem including expected versus actual behavior**:

The implementation of the [C3 algorithm](https://www.usenix.org/system/files/conference/nsdi15/nsdi15-paper-suresh.pdf) uses 1/serviceTime in place of serviceTime.

i.e.
https://github.com/elastic/elasticsearch/blob/fd3d7e336822ed9966cec2be9a4c2d95b438117b/server/src/main/java/org/elasticsearch/node/ResponseCollectorService.java#L176
should be
```
double muBarS = FACTOR / serviceTime;
```

The confusion derives from the fact that the paper in question uses the notation mu_s to denote the **service rate**, which is the inverse of the service-time.

This results in incorrect choices of the fastest node. 

**Steps to reproduce**:

For example, if you apply this change to tests:
```
diff --git a/server/src/test/java/org/elasticsearch/cluster/routing/OperationRoutingTests.java b/server/src/test/java/org/elasticsearch/cluster/routing/OperationRoutingTests.java
index c05d0a551fe..f2bf47e49b2 100644
--- a/server/src/test/java/org/elasticsearch/cluster/routing/OperationRoutingTests.java
+++ b/server/src/test/java/org/elasticsearch/cluster/routing/OperationRoutingTests.java
@@ -571,8 +571,8 @@ public class OperationRoutingTests extends ESTestCase{
         assertThat(searchedShards.size(), equalTo(3));
 
         // Now let's start adding node metrics, since that will affect which node is chosen
-        collector.addNodeStatistics("node_0", 2, TimeValue.timeValueMillis(200).nanos(), TimeValue.timeValueMillis(150).nanos());
-        collector.addNodeStatistics("node_1", 1, TimeValue.timeValueMillis(100).nanos(), TimeValue.timeValueMillis(50).nanos());
+        collector.addNodeStatistics("node_0", 1, TimeValue.timeValueMillis(200).nanos(), TimeValue.timeValueMillis(150).nanos());
+        collector.addNodeStatistics("node_1", 1, TimeValue.timeValueMillis(200).nanos(), TimeValue.timeValueMillis(50).nanos());
         collector.addNodeStatistics("node_2", 1, TimeValue.timeValueMillis(200).nanos(), TimeValue.timeValueMillis(200).nanos());
         outstandingRequests.put("node_0", 1L);
         outstandingRequests.put("node_1", 1L);
```
You can see that with the same number of queued requests and the same client response time for each node, the algorithm should choose the node with the lowest service time (i.e. node 1), however the test fails because the current implementation chooses node_2 (the slowest).

**Provide logs (if relevant)**:

**Comment by pgomulka** [comment]
@robertclancy can you add details how is this related to commit c3ab6f7 ?

**Comment by robertclancy** [comment]
This isn't relevant to https://github.com/elastic/elasticsearch/commit/c3ab6f71c873551ebef333c655b348023a497d9b in particular, that was just the version I happened to test this with.

**Comment by jtibshirani** [comment]
Thank you @robertclancy for raising this. It also looks to me like we've inverted the contribution from service time. I'll follow up with the team to verify, as I am surprised it didn't surface in our benchmarking.

**Comment by jtibshirani** [comment]
I looked into this further and agree that the definition of `muBarS` needs to be updated: it should represent the service rate, not the service time as it does now. I verified the calculation against some example `adaptive_selection` output from `node_stats`.

If we let `R` be average response time, `S` be average service time, and `q` be the queue length estimate, then our current formula can be written as `R + 1 / S * (q^3 - 1). For moderate values of `q`, the `R` term dominates, so we often just rank based on response time.

This can help explain why we see reasonable behavior overall. However, it misses out on important aspects of the algorithm like incorporating non-local statistics, and may cause incorrect selections with larger queue sizes. I think we should fix this and run tests to assess the impact.

**Comment by robertclancy** [comment]
To give more context on how I came across this:

In our production clusters we experience single nodes becoming unresponsive due to, say, hardware failures. In some of these scenarios we see up to 50% of requests fail until the offending node is ejected from the cluster, i.e. after `follower_check.timeout * follower_check.retry_count = 30s` has passed. My understanding was that ARS should help in this scenario but that was not what I was observing.

My belief/hope is that ARS has not been helping as quickly as it should because with the current implementation, as you say, the queue size factor is, effectively, underweighted.

@jtibshirani I'd love to know from your testing whether a change in ARS could help in this scenario (or feel free to correct my, admittedly likely naive, understanding of how this all fits together).

**Comment by jtibshirani** [comment]
@robertclancy sorry for the slow reply. To the best of my knowledge, this fix will not help in that case. Node connection issues do not have an impact on ARS statistics, so the algorithm will not know to adapt and stop sending requests to those nodes. 

However, it seems like an important case we may want to consider. I'll file a new issue if we start to plan improvements and link to it here.

**Comment by robertclancy** [comment]
@jtibshirani I had thought that ARS should help in this case because if the number of outstanding requests in the queue rises, ARS should route away from those nodes.

The reason I expected this issue may be preventing that from happening is because, as you say, the previous implementation had under-weighted the queue-size term because of the inversion of the service time.

**Comment by jtibshirani** [comment]
This is a good point! I hadn't understood your case correctly. If search requests are hanging but not cleanly failing, then this would contribute to the number of outstanding requests, which can increase the ARS rank for that node. To be sure, we would have to test an exact scenario, but from rough experiments it does look like ARS should respond quickly in this case (substantially faster than the follower check timeout).

If you get the chance to try this out once it's released in 7.13, it'd be interesting to hear your feedback. We could move that discussion to the forums (https://discuss.elastic.co/) -- feel free to tag me there if you end up posting.
